lib/Frontend/CompilerInvocation.cpp

```diff
@@ -435,18 +435,28 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
       return true;
     }
   } else if (Opts.RequestedAction != FrontendOptions::NoneAction) {
-    if (Opts.InputFilenames.empty()) {
+    if (Opts.InputFilenames.empty() && !Args.hasArg(OPT_e)) {
       Diags.diagnose(SourceLoc(), diag::error_mode_requires_an_input_file);
       return true;
     }
   }
 
   if (Opts.RequestedAction == FrontendOptions::Immediate) {
-    assert(!Opts.InputFilenames.empty());
-    Opts.ImmediateArgv.push_back(Opts.InputFilenames[0]); // argv[0]
-    if (const Arg *A = Args.getLastArg(OPT__DASH_DASH)) {
-      for (unsigned i = 0, e = A->getNumValues(); i != e; ++i) {
-        Opts.ImmediateArgv.push_back(A->getValue(i));
+    if (Args.hasArg(OPT_e)) {
+      for (const Arg *A : Args.filtered(OPT_e)) {
+        if (A->getNumValues() > 0) {
+          auto sourceBuffer =
```

Comment:
This isn't *quite* right because it will make one buffer per `-e` argument, instead of concatenating them all into a single buffer. Instead, I suggest using `llvm::SmallString` to build up a buffer on the stack and then copy it into an `llvm::MemoryBuffer` once you're done. (SmallString is smart enough to spill onto the heap when it needs to.) Don't forget to insert newlines!